### DISCUSSION TITLE
binding data with x-for

### DISCUSSION THREAD

**Discussion by JonDelWeb** [alpinejs/alpine#896] [1. Help]
Hello guys, 

It's again me ! ;)

I know we are not on StackOverflow but I have a question about .... x-for !!

I would like doing that : 

```
 function loadEmployees() {
            return {
                search: "",
                myForData: sourceData,
                get filteredEmployees() {
                    if (this.search === "") {
                        return this.myForData;
                    }
                    return this.myForData.filter((item) => {
                        return item.employee_name
                            .toLowerCase()
                            .includes(this.search.toLowerCase());
                    });
                },
            };
        }
```

used in : 

```
<div class="container pt-8 mx-auto" x-data="loadEmployees()">
        <input 
            x-ref="searchField" 
            x-model="search" 
            x-on:keydown.window.prevent.slash="$refs.searchField.focus()"
            placeholder="Search for an employee..." 
            type="search"
            class="block w-full bg-gray-200 focus:outline-none focus:bg-white focus:shadow text-gray-700 font-bold rounded-lg px-4 py-3" />
            
        <div class="mt-4 grid grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-4">
            <template x-for="item in filteredEmployees" :key="item">
                <div
                    class="flex items-center shadow hover:bg-indigo-100 hover:shadow-lg hover:rounded transition duration-150 ease-in-out transform hover:scale-105 p-3">
                    <img class="w-10 h-10 rounded-full mr-4" :src="`${item.profile_image}`" />
                    <div class="text-sm">
                        <p class="text-gray-900 leading-none"
                            x-text="item.employee_name + ' (' + item.employee_age + ')'"></p>
                        <p class="text-gray-600" x-text="'$'+item.employee_salary/100"></p>
                    </div>
                </div>
            </template>
        </div>
    </div>
```

BUT where ` sourceData` is fetched data...

How can I put fetched data in arrays and work like this for a searching bar ?

**Comment by pomartel** [comment]
I would initiate `myForData` with an empty array and then fetch the remote data in an `x-init` call. This way the `x-for` won't break when it tries to first render with no data.

**Reply by JonDelWeb** [reply]
Thanks a lot... 

Unfortunately and I don't know why, that's don't work if I use destructuring data like this : 

```
<body x-data="{...application(), ...connections()}" x-init="getListApps()">
```

and my js : 

```
function application() {
    return {
        openTab: 1,
        search: "",
        list: [],
        getListApps: async function () {
                const response = await fetch('https://my-json-server.typicode.com/JonDelWeb/appList/application');
                const data = await response.json();
                this.list = data;
        },
        get listApps() {
            if (this.search === "") {
                return this.list;
            }
            return this.list.filter((item) => {
                return item.name
                    .toLowerCase()
                    .includes(this.search.toLowerCase());
            });
        },
    }
}
```
and I have the connections() function in a other files...

**Reply by pomartel** [reply]
And it works fine without the destructuring? Do you maybe have the same function name in both objects?

**Reply by JonDelWeb** [reply]
Yes, if I use like this : 

```
<body x-data="main()">

<div x-data="application()" x-init="getListApps()">
       .... some other content
<div>

<div x-data="connections()" x-init="getListApps()">
  .... some other content
<div>

</body>
```

that's work... but I must have the getListApps in the both x-data component and it's no optimum for me...

**Comment by pomartel** [comment]
Maybe you could try writing a little proof of concept Codepen so it would
be easier to understand where the problem could be:

https://codepen.io/pen?template=bGdzxBO

**Comment by JonDelWeb** [comment]
I do it : 

https://codepen.io/JonDelWeb/pen/wvzaMGo

I can't use the search bar... And I would like to put it in the top of body...

**Comment by pomartel** [comment]
Both your Application and Connections objects have the same function names.
So the connections functions will just override the application functions
with the same names. You should try giving your keys different names in
both objects or prefix them.

**Reply by JonDelWeb** [reply]
So I'm really sorry but I don't understand how it's work...

**Comment by JonDelWeb** [comment]
Hello, it's again me...

I really want understand how its work under the hood but...

I change my code like this : 

https://codepen.io/JonDelWeb/pen/wvzaMGo

But the search bar still doesn't work...
I tried a lot of things but nothing works, I must have missed something...

**Reply by SimoTod** [reply]
your endpoint is broken. It returns a 507 response `{error: "maxContentLength size of 10000 exceeded"}`

**Reply by JonDelWeb** [reply]
I fix it ! ;)

**Reply by SimoTod** [reply]
There are a few typos. your model sometimes is called search and sometimes is called searchApp.
Using spread with a getter is not a good idea because it will keep the scope of the original object.
Your getter is called listApps but, in your template, you use listApp, etc etc.

See https://codepen.io/SimoTod/pen/mdrJzYL

**Reply by JonDelWeb** [reply]
Ok thanks a lot... I understand better I think. But I prefer another approach where I separate the components well. 

Thanks alot again and sorry for the inconvenience.
